### ISSUE TITLE
Feature: Hashicorp Vault as additional secret store

### ISSUE THREAD

**Issue by mkretz** [Azure/azure-service-operator#2326]
We are using Hashicorp Vault as our central secret store, and we would like ASO to store and retrieve credentials there instead of in Kubernetes secrets or in Azure Key Vault. Therefore, we are considering creating a PR for ASO v1 which 

1. creates a new implementation of [SecretClient](https://github.com/Azure/azure-service-operator/blob/96034362a35565a808f253655f328452e45ce683/pkg/secrets/interface.go#L22) for Hashicorp Vault and
2. which extends the ASO configuration and custom resources by suitable attributes for specifying the Hashicorp Vault to be used.

Hence, we have the following question (before we start investing time): would such a PR be welcome?

**Comment by matthchr** [comment]
Sorry for the slow response @mkretz.

As you're probably aware, ASOv1 is in maintenance mode. That means we're fixing issues but generally avoiding new features. We only support Kubernetes secrets in ASOv1, and it's done in a slightly different mechanism. We were hoping that need for alternative secret stores was decreasing with managed identity being supported by more and more resources.

Can you tell us more about what CRDs you're using that produce/consume secrets which you would like to move to Vault? What are you using for secret storage now, k8s secrets or KeyVault? 

I believe we would lean towards not accepting any feature additions to ASOv1, but if we knew more about your scenarios and the resources you're using we could comment on the possibility of supporting this in ASOv2.

**Comment by mkretz** [comment]
No problem @matthchr! Thanks for getting back to me!

Sure! Here's an outline of our current setup. Short version up front: we have found a way to proceed without a PR. 

- We have one central Azure SQL Server which we set up using Terraform (i.e. not via ASO). The admin credentials of this Azure SQL Server are written to Azure Key Vault and also reflected to a Kubernetes secret using the standard integration for Azure Key Vault and Kubernetes. 
- We then have a Helm chart that sets up several `azuresqldatabase` resources and along with them `azuresqluser` resources.
- `azuresqldatabase` resources are set up by ASO using Managed Identity, so no need for Vault there, as you mention.
- For the `azuresqluser` resources, ASO needs the admin credentials for the Azure SQL server, which in our case it gets from the Kubernetes secret mentioned above (could also get it directly from Azure Key Vault, but we need the Kubernetes secrets for something else too)
- Our applications then access their respective databases by reading the credentials from the corresponding `azuresqluser` secret into env variables.

Since certain parts of our architecture are not in Azure we rely on Hashicorp Vault as a central secrets store, so our initial reflex was that it would be necessary to be able to configure Hashicorp Vault as the secret store in particular in `azuresqluser`. However, after having done some additional analysis, we have decided that it will be better in our case to use the [Hashicorp Vault Databases Engine](https://www.vaultproject.io/docs/secrets/databases) which will take care of user creation and credential rotation. Hence, we will no longer use the `azuresqluser` resource. This also removes our need for a PR for now.

We can close this issue. Thanks again!
